fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,809 @@
+
+/// The variable dispatch table. This is set at startup and cannot be modified after.
+static VAR_DISPATCH_TABLE: once_cell::sync::Lazy<VarDispatchTable> =
+    once_cell::sync::Lazy::new(|| {
+        let mut table = VarDispatchTable::default();
+
+        for name in LOCALE_VARIABLES {
+            table.add_anon(name, handle_locale_change);
+        }
+
+        for name in CURSES_VARIABLES {
+            table.add_anon(name, handle_curses_change);
+        }
+
+        table.add(L!("TZ"), handle_tz_change);
+        table.add_anon(L!("fish_term256"), handle_fish_term_change);
+        table.add_anon(L!("fish_term24bit"), handle_fish_term_change);
+        table.add_anon(L!("fish_escape_delay_ms"), update_wait_on_escape_ms);
+        table.add_anon(L!("fish_emoji_width"), guess_emoji_width);
+        table.add_anon(L!("fish_ambiguous_width"), handle_change_ambiguous_width);
+        table.add_anon(L!("LINES"), handle_term_size_change);
+        table.add_anon(L!("COLUMNS"), handle_term_size_change);
+        table.add_anon(L!("fish_complete_path"), handle_complete_path_change);
+        table.add_anon(L!("fish_function_path"), handle_function_path_change);
+        table.add_anon(L!("fish_read_limit"), handle_read_limit_change);
+        table.add_anon(L!("fish_history"), handle_fish_history_change);
+        table.add_anon(
+            L!("fish_autosuggestion_enabled"),
+            handle_autosuggestion_change,
+        );
+        table.add_anon(
+            L!("fish_use_posix_spawn"),
+            handle_fish_use_posix_spawn_change,
+        );
+        table.add_anon(L!("fish_trace"), handle_fish_trace);
+        table.add_anon(
+            L!("fish_cursor_selection_mode"),
+            handle_fish_cursor_selection_mode_change,
+        );
+
+        table
+    });
+
+type NamedEnvCallback = fn(name: &wstr, env: &dyn Environment);
+type AnonEnvCallback = fn(env: &dyn Environment);
+
+#[derive(Default)]
+struct VarDispatchTable {
+    named_table: HashMap<&'static wstr, NamedEnvCallback>,
+    anon_table: HashMap<&'static wstr, AnonEnvCallback>,
+}
+
+// TODO: Delete this after input_common is ported (and pass the input_function function directly).
+fn update_wait_on_escape_ms(vars: &dyn Environment) {
+    let fish_escape_delay_ms = vars.get_unless_empty(L!("fish_escape_delay_ms"));
+    let var = crate::env::environment::env_var_to_ffi(fish_escape_delay_ms);
+    crate::ffi::update_wait_on_escape_ms_ffi(var);
+}
+
+impl VarDispatchTable {
+    fn observes_var(&self, name: &wstr) -> bool {
+        self.named_table.contains_key(name) || self.anon_table.contains_key(name)
+    }
+
+    /// Add a callback for the variable `name`. We must not already be observing this variable.
+    pub fn add(&mut self, name: &'static wstr, callback: NamedEnvCallback) {
+        let prev = self.named_table.insert(name, callback);
+        assert!(prev.is_none(), "Already observing {}", name);
```

Comment:
The previous code checked `observes_var` which looks in both tables. It's not super-important but let's preserve the existing behavior.

Reply:
That was an oversight on my part. Thanks for catching it.